File: scala/src/test/scala/tddmicroexercises/leaderboard/LeaderboardSpec.scala
package tddmicroexercises.leaderboard

import org.scalatest.{Matchers, FlatSpec}

class LeaderboardSpec extends FlatSpec with Matchers {

  it should "sum the points" in {
    TestData.sampleLeaderboard1.driver_results()("Lewis Hamilton") should equal(18+18+25)
  }

  it should "calculate the winner" in {
    TestData.sampleLeaderboard1.driver_rankings()(0) should equal ("Lewis Hamilton")
  }
}

class RaceSpec extends FlatSpec with Matchers {
  it should "calculate driver points" in {
    TestData.race1.points(TestData.driver1) should equal(25)
    TestData.race1.points(TestData.driver2) should equal(18)
    TestData.race1.points(TestData.driver3) should equal(15)
  }
}

/**
 * Test Data found via http://en.wikipedia.org/wiki/2015_Formula_One_season
 */
object TestData {

  val driver1 = new Driver(name="Nico Rosberg", country="DE")
  val driver2 = new Driver(name="Lewis Hamilton", country="UK")
  val driver3 = new Driver(name="Sebastian Vettel", country="DE")
  val driver4 = new SelfDrivingCar(algorithmVersion = "1.2", company="Acme")

  val race1 = new Race("Australian Grand Prix", List(driver1, driver2, driver3))
  val race2 = new Race("Malaysian Grand Prix", List(driver3, driver2, driver1))
  val race3 = new Race("Chinese Grand Prix", List(driver2, driver1, driver3))
  val race4 = new Race("Fictional Grand Prix 1", List(driver1, driver2, driver4))
  val race5 = new Race("Fictional Grand Prix 2", List(driver4, driver2, driver1))
  driver4.algorithmVersion = "1.3"
  val race6 = new Race("Fictional Grand Prix 3", List(driver2, driver1, driver4))

  val sampleLeaderboard1 = new Leaderboard(List(race1, race2, race3))
  val sampleLeaderboard2 = new Leaderboard(List(race4, race5, race6))
}

File: Ruby/LeaderBoard/spec/leader_board_spec.rb
require_relative '../leader_board'
describe LeaderBoard do
  before :each do
    @driver_1 = Driver.new("Nico Rosberg", "DE")
    @driver_2 = Driver.new("Lewis Hamilton", "UK")
    @driver_3 = Driver.new("Sebastian Vettel", "DE")
    @driver_4 = SelfDrivingCar.new("1.2", "ACME")

    @race_1 = Race.new("Australian Grand Prix", [@driver_1, @driver_2, @driver_3])
    @race_2 = Race.new("Malaysian Grand Prix", [@driver_3, @driver_2, @driver_1])
    @race_3 = Race.new("Chinese Grand Prix", [@driver_2, @driver_1, @driver_3])
    @race_4 = Race.new("Fictional Grand Prix", [@driver_1, @driver_2, @driver_4])
    @race_5 = Race.new("Fictional Grand Prix", [@driver_4, @driver_2, @driver_1])
    @driver_4.algorithm_version = "1.3"
    @race_6 = Race.new("Fictional Grand Prix", [@driver_2, @driver_1, @driver_4])

    @sample_leader_board_1 = LeaderBoard.new([@race_1, @race_2, @race_3])
    @sample_leader_board_2 = LeaderBoard.new([@race_4, @race_5, @race_6])
  end
  
  context "#winner" do
    it "returns the driver with the most points" do
      expect(@sample_leader_board_1.driver_rankings.first).to eq("Lewis Hamilton")
    end
  end

  context "#driver_points" do
    it "returns the correct number of points for a driver" do
      expect(@sample_leader_board_2.driver_points["Lewis Hamilton"]).to eq(18 + 18 + 25)
    end
  end
end
